### DISCUSSION TITLE
'Is' pattern in pointers

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#4423] [Q&A] [answered]
Hello.

C# allows `is` pattern on pointers, but I can't figure the differences out between `pointer == null` and `pointer is null`. I wonder that what real differences between those two kinds of the code style of comparison. Or in other words, why I should use `pointer is null` instead of `pointer == null` or vice versa?

Test code:

```csharp
using System;

unsafe
{
    int* p = null;
    delegate*<void> q = null;

    if (p == null) Console.WriteLine(1);
    if (p is null) Console.WriteLine(1);

    if (p != null) Console.WriteLine(1);
    if (p is not null) Console.WriteLine(1);

    if (q == null) Console.WriteLine(1);
    if (q is null) Console.WriteLine(1);

    if (q != null) Console.WriteLine(1);
    if (q is not null) Console.WriteLine(1);
}
```

[See this link (SharpLab)](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKFwVwDsoIAzOXAb1wAI6aBLQmAKhoAcaBeGw/AG34BuWvQAmcfnADmEeCwA8KACwA+GgEduvAcNyi6DUjQAUnLjz6CAlDRQYAnCYzWROeo2NnGUHTbuOzq767vRGppwAhJa6tvZOLm4e4d4MvoQA9jB+/HGBiSHJXloWOXkJwaGGxT5lARVuBp6mWtF18UFJYTVpvFnt+ZUAvkA==)

Thanks.

**Comment by huoyaoyuan** [comment, marked as answer]
I believe the real reason is syntax matrix completeness, because we don't have a reason to ban it.
Allowing pointers to be handled by pattern matching is useful for nested cases like `is { PointerProp: not null }`.

**Reply by kyoyama-kazusa** [reply]
Oh, that sounds reasonable! Thank you!
